Fixing Exponential Quest Reward Scrolls

Hus274 opened the pull request:
Fixes #2715 
### Changes

Updating Recidivate, Golden Knight, Attack of the Mundane, and Vice quests to only reward the next quest scroll in the story line to the quest owner. This should remove the exponential awarding of quest scrolls in a party with multiple people.

Hover Over Quest in Market:
![fullpicture](https://cloud.githubusercontent.com/assets/17734753/16901553/103f6a2a-4bfc-11e6-9970-6fa42c2142f5.png)

Quest Start:
![queststart](https://cloud.githubusercontent.com/assets/17734753/16901558/2268fb30-4bfc-11e6-9dc9-6cd325a67f93.png)

Quest Completed:
![questcompleted](https://cloud.githubusercontent.com/assets/17734753/16901561/27a3b496-4bfc-11e6-936b-d90c3610e71a.png)

---

UUID: 2a4d30ee-4833-420e-be7c-ad6a6e261c5f

Hus274 committed:
adding quest owner specific rewards. closes #2715

paglias commented on common/script/content/index.js:
```diff
@@ -1251,7 +1251,7 @@ api.quests = {
     },
     drop: {
-      items: [
+      itemsForOwner: [
```
why this name change? It's probably a breaking change that we cannot make without breaking mobile apps

Hus274 replied:
The idea was to split the awards of a quest into two different arrays, items for all participants and items for just the quest owner. It's not a rename so much as moving some of the rewards into a new array titled itemsForOwner. 

That being said, it probably will be a breaking change in terms of correctly displaying the quest object on the mobile app. The only other way I can think of distinguishing these is to update the item model and add another field to it that says who it applies to. So something like item.awardsOnlyQuestOwner. The reason I didn't do this in the first place was because it seemed to be less efficient as then getting the list of rewards for allParticipants and for just the quest owner would both require a mapping function on the full list to turn it into the two arrays.

If this would be a breaking change, it's probably worth sacrificing the efficiency in favor of not breaking the apps. Should I update this to the item property model?

paglias replied:
Yeah having `item.onlyOwner` or something similar seems the best solution. These are small arrays and it's not going to be a perf problem having to filter them

Hus274 replied:
Makes sense to me! I'll update this and resubmit

Hus274 committed:
Updating model to prevent this from being a breaking change.

Hus274 commented:
Updated the model so this won't be a breaking change.

khipkin commented on common/locales/en/quests.json:
```diff
@@ -10,7 +10,13 @@
     "completed": "Completed!",
+    "rewardsAllParticipants": "Rewards for all Quest Participants",
+    "rewardsQuestOwner": "Additional Rewards for Quest Owner",
+    "questOwnerReceived": "The Quest Owner has also Received",
```
You have this string twice.

Hus274 replied:
Good catch, I was doing some pretty interesting git stash stuff and I must have missed this on one of the pops. I'll fix this when the rest is reviewed.

paglias replied:
Is Received meant to start with an upper case letter?

Hus274 replied:
Originally they were all capitalized as I based them on the image in #2715. I can update this to be lowercase though

crookedneighbor replied:
It's supposed to be capitalized as it is a title.

Hus274 replied:
Should it be "Quest Owner Has Also Received"?

khipkin commented on website/views/shared/modals/quest-rewards.jade:
```diff
@@ -1,8 +1,8 @@
+  h5 {{::headerParticipant}}
   table.table.table-striped
-    tr(ng-repeat='drop in quest.drop.items')
+    tr(ng-repeat='_.reject(quest.drop.items, \'onlyOwner\')')
```
no more `drop in`?

Hus274 committed:
Removing duplicate translatable string and readding accidentally deleted portion

paglias commented:
Looking good to me

Hus274 committed:
capitalizing according to pr.

Hus274 commented:
Capitalization added to the title

Hus274 committed:
Merging into develop

Hus274 commented:
Rebased on top of develop and the tests are passing again

crookedneighbor commented on test/api/v3/unit/models/group.test.js:
```diff
@@ -929,9 +929,19 @@ describe('Group Model', () => {
 
```
Let's specify in the it block that it awards quest scrolls to the owner.

crookedneighbor commented on website/client/js/services/questServices.js:
```diff
@@ -68,9 +68,10 @@ angular.module('habitrpg')
-      if(quest.drop.items) {
-        for (var item in quest.drop.items) {
+      text += '**' + window.env.t('rewardsAllParticipants') + ':**\n\n';
+      var participantRewards = quest.drop.items ? _.reject(quest.drop.items, 'onlyOwner') : quest.drop.items;
```
I think:

``` js
var participantRewards = quest.drop.items ? _.reject(quest.drop.items, 'onlyOwner') : null;
```

Would have a little more clarity here.

crookedneighbor replied:
Also, I believe `_.reject` works with falsy values:

``` js
var participantRewards = _.reject(undefined, 'onlyOwner') // []
if (particpantRewards.length > 0) {
  // logic
}
```

Double check that, but it could make the logic more readable.

crookedneighbor commented on website/client/js/services/questServices.js:
```diff
@@ -68,9 +68,10 @@ angular.module('habitrpg')
+      text += '**' + window.env.t('rewardsAllParticipants') + ':**\n\n';
+      var participantRewards = quest.drop.items ? _.reject(quest.drop.items, 'onlyOwner') : quest.drop.items;
+      if(participantRewards) {
+        for (var item in participantRewards) {
```
Let's use `_.each` here. `for in` loops have some bad behavior when the Object.prototype is modified by a third party. (I know you didn't write the original `for in`, but we should still clean it up.

crookedneighbor commented on website/client/js/services/questServices.js:
```diff
@@ -79,6 +80,14 @@ angular.module('habitrpg')
       if(quest.drop.gp)
         text += quest.drop.gp + ' ' + window.env.t('gold') + '\n\n';
 
+      var ownerRewards = quest.drop.items ? _.filter(quest.drop.items, 'onlyOwner') : quest.drop.items;
```
same comments as above

Commit:
Merge remote-tracking branch 'upstream/develop' into 2715

Commit:
fixing according to comments on pr

Commit:
removing final mistakes

crookedneighbor commented on website/client-old/js/services/questServices.js:
```diff
@@ -68,17 +68,26 @@ angular.module('habitrpg')
+      text += '**' + window.env.t('rewardsAllParticipants') + ':**\n\n';
+      var participantRewards = _.reject(quest.drop.items, 'onlyOwner');
+      if(participantRewards.length > 0) {
+        _.each( participantRewards, function(item) {
```
remove the space after `_.each(`, so that it is `_.each(participantRewards`

Hus274 replied:
done.

crookedneighbor commented on website/client-old/js/services/questServices.js:
```diff
@@ -68,17 +68,26 @@ angular.module('habitrpg')
+      var ownerRewards = _.filter(quest.drop.items, 'onlyOwner');
+      if(ownerRewards.length > 0) {
+        text += '**' + window.env.t('rewardsQuestOwner') + ':**\n\n';
+        _.each( ownerRewards, function(item){
```
Same here.

Hus274 replied:
done

Hus274 commented:
Finished up fixing this according to the comments here. Also tested the _.reject(undefined, ...) behavior to make sure that it worked as @crookedneighbor said and everything looks good!

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -507,38 +539,33 @@ schema.methods.finishQuest = async function finishQuest (quest) {
+  _.each(_.filter(quest.drop.items, 'onlyOwner'), (item) => {
+    _.merge(questOwnerUpdates, _getUserUpdateForQuestReward(item, quest.drop.items));
   });
+  questOwnerUpdates = _.omit(questOwnerUpdates, _.isEmpty);
```
I feel like this should be done in the `_getUserUpdateForQuestReward`  function.

Hus274 replied:
I'm not sure I understand. The method right now creates the update object that gets pushed to all users participating in the quest and then comes back and generates the update specifically for the owner of the quest.

What specifically did you want moved into the method above and what would the return value be?

crookedneighbor replied:
Maybe I don't understand what this is doing. I thought it was stripping out any update key that was an empty object. IE, if `$set` never gets a value, it removes it. Is there any reason to do it out here rather than in the function itself?

https://runkit.com/crookedneighbor/lodash-3-omit-test

Hus274 replied:
Ah gotcha, that totally makes sense. When I was looking at this initially, i thought this was a comment on a block of code not on this single line which is why I didn't understand. What you suggested makes perfect sense, my mistake. I've updated the pull request accordingly.

Commit:
Merge remote-tracking branch 'upstream/develop' into 2715

Hus274 reviewed (commented):
Addressed comments and left a question on one

Commit:
fixing whitespace

Commit:
Merge remote-tracking branch 'upstream/develop' into 2715

Commit:
re-adding the onlyOwner field that got deleted when the index.js file was moved and fixed console errors.

Hus274 commented:
Re-added the missing only owner field and fixed console errors.

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 2715

Hus274 committed:
moving cleaning of empty obejct for quest owner updates into quest owner updates method

Hus274 commented:
Added change to move omit into the method.

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 2715

Hus274 committed:
Fixing so tests pass by updating variable name and removing unnecessary parameter definition.

Hus274 commented:
Sorry, I totally missed that these were failing the test suite. I've made the final adjustments to get them to pass.

This should be totally ready to go now! Let me know if there's anything else you need me to change here.

TheHollidayInn commented on website/views/shared/modals/quest-rewards.jade:
```diff
@@ -12,3 +12,8 @@ script(id='partials/options.social.party.quest-rewards.html', type='text/ng-temp
         =env.t('gold')
     tr(ng-if='::quest.drop.unlock()')
       td {{::quest.drop.unlock()}}
+  div(ng-if='_.filter(quest.drop.items, \'onlyOwner\').length > 0')
```
One last comment for this PR. Can you move `_.filter(quest.drop.items, \'onlyOwner\')` to a variable on the controller. OR, since this is a model, move it to the Modal scope?

Hus274 replied:
Yep, no problem. I'll try to be better about keeping this sort of logic out of the display layer in the future.

TheHollidayInn replied:
Sounds good. I'm trying to get your older PRs merged in, so no worries :P

Hus274 replied:
I really appreciate that! Just trying to do my part to make your job easier! 😄

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 2715

Hus274 committed:
adding a new test and refactoring client side code to use controller method.

Hus274 commented:
I've fixed the merge conflicts, added a new test case, and moved the code on the client side into a controller method. Let me know if there's anything else needed here

TheHollidayInn commented:
Noted towards your tier!
